listener/upload.go

```diff
@@ -0,0 +1,153 @@
+package listener
+
+import (
+	"app/base/database"
+	"app/base/models"
+	"app/base/utils"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"github.com/RedHatInsights/patchman-clients/inventory"
+	"github.com/RedHatInsights/patchman-clients/vmaas"
+	"github.com/antihax/optional"
+	"github.com/segmentio/kafka-go"
+	"time"
+)
+
+func uploadHandler(m kafka.Message) {
+	var event PlatformEvent
+	utils.Log("msg", string(m.Value)).Info("Msg received")
+
+	err := json.Unmarshal(m.Value, &event)
+	if err != nil {
+		utils.Log("err", err.Error()).Error("Could not deserialize host event")
+		return
+	}
+	// We need the b64 identity in order to call the inventory
+	if event.B64Identity == nil {
+		utils.Log().Error("No identity provided")
+		return
+	}
+
+	identity, err := utils.ParseIdentity(*event.B64Identity)
+	if err != nil {
+		utils.Log("err", err.Error()).Error("Could not parse identity")
+		return
+	}
+
+	if !identity.IsSmartEntitled() {
+		utils.Log("account", identity.Identity.AccountNumber).Info("Is not smart entitled")
+		return
+	}
+	hostUploadReceived(event.Id, identity.Identity.AccountNumber, *event.B64Identity)
+}
+
+// Stores or updates the account data, returning the account id
+func getOrCreateAccount(account string) (int, error) {
+	rhAccount := models.RhAccount{
+		Name: account,
+	}
+
+	err := database.OnConflictUpdate(database.Db, "name", "name").Create(&rhAccount).Error
+	return rhAccount.ID, err
+}
+
+// Stores or updates base system profile, returing internal system id
+func updateSystemPlatform(inventoryId string, accountId int, updatesReq *vmaas.UpdatesRequest) (int, error) {
+	updatesReqJSON, err := json.Marshal(&updatesReq)
+	if err != nil {
+		utils.Log("err", err.Error()).Error("Serializing vmaas request")
+		return 0, err
+	}
+
+	hash := sha256.New()
+	hash.Write(updatesReqJSON)
+	jsonChecksum := hex.EncodeToString(hash.Sum([]byte{}))
+
+	now := time.Now()
+
+	dbHost := models.SystemPlatform{
+		InventoryID:    inventoryId,
+		RhAccountID:    accountId,
+		FirstReported:  &now,
```

Comment:
Is it sure, that this is the first upload?

Reply:
This field is set by database trigger. Fixed.